### PR TITLE
[misc] refactor: re-format examples and deprecate old examples

### PR DESCRIPTION
**Pull request by ETOgaosion**
### What does this PR do?

Current examples are too heavy, some old ones may need to patch or modify verl's main-stream code and already placed in recipe, clear useless and repeated examples with feature tests. We derectly reformat all tests naming and features using common practices.

# verl Examples

Prefer include scripts-only examples, meaning that your example can directly use main-stream verl's feature

## Conventions

All run scripts follow the same shape:

1. Canonical filename:

   ```
   run_<model>_<train-backend>.sh
   ```

   - `<model>`: a single canonical size per model family. E.g.
     `qwen3_8b`, `qwen3_30b_a3b`, `qwen3_235b_a22b`, `qwen3_vl_8b`,
     `deepseek_v3`, `mimo_7b`, `nemotron_nano_v3`.
   - `<train-backend>`: `fsdp`, `fsdp2`, `megatron`, or `mindspeed`.
   - Now that some scripts directly support both NPU and GPU running, just use DEVICE variable to control (options: ['npu', 'cuda'])

   Per-example *features* (Liger kernel, LoRA, sequence parallel size, server
   vs sync rollout, etc.) do **not** show up in filenames — they are exposed as
   env-var toggles inside the one canonical script. For example, `sft/gsm8k/
   run_qwen2_5_0_5b_fsdp.sh` covers plain SFT and its `USE_LIGER=1`, `SP_SIZE=2`,
   `USE_PEFT=1` variants via env vars.

2. Every script exposes its important knobs as env vars at the top, e.g.

   ```bash
   MODEL_PATH=${MODEL_PATH:-Qwen/Qwen3-8B}
   NNODES=${NNODES:-1}
   NGPUS_PER_NODE=${NGPUS_PER_NODE:-8}
   TRAIN_BATCH_SIZE=${TRAIN_BATCH_SIZE:-1024}
   ROLLOUT_TP=${ROLLOUT_TP:-2}
   ROLLOUT_N=${ROLLOUT_N:-5}
   ...
   ```

   Override anything you care about on the command line:

   ```bash
   MODEL_PATH=/my/local/qwen3-8b NGPUS_PER_NODE=4 bash examples/grpo_trainer/run_qwen3_8b_vllm_fsdp.sh
   ```

3. Defaults (unless a directory explicitly documents otherwise):

   - `data.train_files` + `data.val_files` = GSM8K + MATH for text LLMs
     (`geo3k` for vision, `dapo-math-17k` / `aime-2024` for scale-demo 235B /
     671B scripts).
   - `actor_rollout_ref.actor.use_dynamic_bsz=True`
   - `trainer.balance_batch=True`
   - `trainer.logger=["console","wandb"]`.

4. No deprecated Hydra knobs:

   - `ppo_megatron_trainer.yaml` → use `actor_rollout_ref.actor.model_engine=megatron`.
   - `ppo_micro_batch_size` / `log_prob_micro_batch_size` → use the
     `_per_gpu` suffix.
   - `data.val_batch_size` → removed.
   - Top-level `reward_model.*` → use `reward_model.reward_model.*` /
     `reward.reward_model.*` as applicable.
   - `actor.ulysses_sequence_parallel_size` → use
     `actor_rollout_ref.actor.fsdp_config.ulysses_sequence_parallel_size`.

### Algorithm trainers

Each directory holds a canonical recipe for one training algorithm. Adding a
new algorithm? If it needs its own trainer entry point or reward code, put it
under `recipe/` instead.

| Dir                                | Algorithm                      | `algorithm.adv_estimator` / `policy_loss.loss_mode` |
|------------------------------------|--------------------------------|-----------------------------------------------------|
| `ppo_trainer/`                     | PPO (actor + critic)           | `adv_estimator=gae`                                 |
| `grpo_trainer/`                    | GRPO                           | `adv_estimator=grpo`                                |
| `rloo_trainer/`                    | RLOO                           | `adv_estimator=rloo`                                |
| `remax_trainer/`                   | ReMax                          | `adv_estimator=remax`                               |
| `reinforce_plus_plus_trainer/`     | REINFORCE++ / baseline         | `adv_estimator=reinforce_plus_plus[_baseline]`      |
| `cispo_trainer/`                   | CISPO                          | `loss_mode=cispo`                                   |
| `dppo_trainer/`                    | DPPO (TV / KL variants)        | `loss_mode=dppo_tv \| dppo_kl`                      |
| `gdpo_trainer/`                    | GDPO                           | `adv_estimator=gdpo`                                |
| `gmpo_trainer/`                    | GMPO                           | `loss_mode=geo_mean`                                |
| `gpg_trainer/`                     | GPG                            | `adv_estimator=gpg`, `loss_mode=gpg`                |
| `gspo_trainer/`                    | GSPO                           | `loss_mode=gspo`                                    |
| `sapo_trainer/`                    | SAPO                           | `loss_mode=sapo`                                    |
| `otb_trainer/`                     | OTB                            | `adv_estimator=optimal_token_baseline`              |
| `mtp_trainer/`                     | DAPO + MTP (MiMo-7B)           | `adv_estimator=grpo`, MTP flags                     |
| `on_policy_distillation_trainer/`  | on-policy distillation         | GRPO + distillation loss                            |
| `flowgrpo_trainer/`                | Flow-GRPO (diffusion)          | image-gen specific                                  |

### Feature / infra

| Dir                  | Purpose                                                                                  |
|----------------------|------------------------------------------------------------------------------------------|
| `tuning/`            | LoRA (`tuning/lora/`) and scaling demos (`tuning/scaling/`).                             |
| `profile/`           | NPU profiler / torch-memory profiler runs.                                               |
| `sft/`               | Supervised fine-tuning examples.                                                         |
| `generation/`        | Rollout-only inference launches.                                                         |
| `sglang_multiturn/`  | SGLang multi-turn rollout examples.                                                      |
| `vllm_omni/`         | vLLM omni backend examples.                                                              |
| `data_preprocess/`   | Scripts that produce the `$HOME/data/<dataset>/*.parquet` layout the run scripts expect. |
| `prefix_grouper/`    | Prefix-grouped rollout examples.                                                         |
| `rollout_correction/`| Rollout correction examples.                                                             |
| `router_replay/`     | Router replay examples.                                                                  |
| `tutorial/`          | Tutorials and cluster launchers (`ray/`, `slurm/`, `skypilot/`, `agent_loop_get_started/`). |

### Where are the algorithm research variants?

`recipe/` — e.g. `recipe/dapo`, `recipe/prime`, `recipe/retool`,
`recipe/r1`, `recipe/spin`, `recipe/gvpo`, `recipe/flowrl`, ...
They ship their own trainer entry points and reward code.

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `veomni`, `sglang`, `vllm`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `fully_async`, `one_step_off`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)
- [ ] If your PR is related to the `recipe` submodule, please also update the reference to the submodule commit via `git submodule update --remote` or `cd recipe && git pull origin main`.

### CODE REVIEW

**Comment by ETOgaosion** [comment]
@wucong25 Please help review~

**Comment by wuxibin89** [comment]
> `run_<model>_<infer-backend>_<train-backend>.sh`

I think `<infer-backend>` can be also omitted, since rollout backend should be seamlessly changed by `actor_rollout_ref.rollout.name`

**Comment by wuxibin89** [comment]
Should we also add a pre-commit check rule for script naming conventions?

File: docs/ascend_tutorial/examples/ascend_sglang_best_practices.rst [hunk 1]
```diff
@@ -3,8 +3,8 @@ Ascend SGLang Best Practice
 
 Last updated: 01/27/2026.
 
-.. _Qwen3-30B: https://github.com/verl-project/verl/blob/main/examples/grpo_trainer/run_qwen3moe-30b_sglang_megatron_npu.sh
-.. _Qwen2.5-32B: https://github.com/verl-project/verl/blob/main/examples/grpo_trainer/run_qwen2-32b_sglang_fsdp_npu.sh
+.. _Qwen3-30B: https://github.com/verl-project/verl/blob/main/examples/grpo_trainer/run_qwen3_30b_a3b_sglang_mindspeed_npu.sh
+.. _Qwen2.5-32B: https://github.com/verl-project/verl/blob/main/examples/grpo_trainer/run_qwen3_8b_sglang_fsdp_npu.sh
```
Comment (hustmf) [inline docs/ascend_tutorial/examples/ascend_sglang_best_practices.rst]: 32B链接了一个8B的脚本吗

**Comment by wucong25** [comment]
Can we not modify the NPU-specific script for now? In the next few days, we will uniformly adjust the script to align with the GPU according to the script submission specifications.

File: docs/ascend_tutorial/examples/ascend_sglang_best_practices.rst (hunk 1 above)
Comment (ETOgaosion) [reply docs/ascend_tutorial/examples/ascend_sglang_best_practices.rst]: This is an error

**Comment by tardis-key** [comment]
Will we test all modified scripts to verify availability?

**Comment by ETOgaosion** [comment]
> Can we not modify the NPU-specific script for now? In the next few days, we will uniformly adjust the script to align with the GPU according to the script submission specifications.

Hello @wucong25 , current refactor just works the same as main stream, there is need for check and supplement all tests to use npu conviniently, feel free to follow the new canonical to help commit.

**Comment by ETOgaosion** [comment]
> Will we test all modified scripts to verify availability?

There is need for tests on different types of machines, we are calling community helps to verify and fix configurations on all examples with different machines. Since the example mainly for runnable cases, we are not asking for adaptation for different machines and best performance.

File: examples/ppo_trainer/run_qwen3_8b_fsdp.sh [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    npu)
+        export HCCL_CONNECT_TIMEOUT=2400
+        export HCCL_EXEC_TIMEOUT=2400
+        export HCCL_OP_EXPANSION_MODE=AIV
+        export CLOSE_MATMUL_K_SHIFT=1
+
+        n_devices_per_node=${NDEVICES_PER_NODE:-${NPUS_PER_NODE:-${NGPUS_PER_NODE:-8}}}
+        project_name=${PROJECT_NAME:-verl_ppo_gsm8k_math_npu}
+        experiment_name=${EXPERIMENT_NAME:-qwen3_8b_${INFER_BACKEND}_fsdp_npu}
+        device_trainer_args+=("trainer.device=npu")
```
Comment (wuxibin89) [inline examples/ppo_trainer/run_qwen3_8b_fsdp.sh]: User should not set device explictly, it's auto set: https://github.com/verl-project/verl/blob/main/verl/trainer/main_ppo.py#L44

File: examples/ppo_trainer/run_qwen3_8b_fsdp.sh [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        project_name=${PROJECT_NAME:-verl_ppo_gsm8k_math}
+        experiment_name=${EXPERIMENT_NAME:-qwen3_8b_${INFER_BACKEND}_fsdp}
+        ;;
+    npu)
+        export HCCL_CONNECT_TIMEOUT=2400
+        export HCCL_EXEC_TIMEOUT=2400
+        export HCCL_OP_EXPANSION_MODE=AIV
+        export CLOSE_MATMUL_K_SHIFT=1
+
+        n_devices_per_node=${NDEVICES_PER_NODE:-${NPUS_PER_NODE:-${NGPUS_PER_NODE:-8}}}
```
Comment (wuxibin89) [inline examples/ppo_trainer/run_qwen3_8b_fsdp.sh]: project_name, experiment_name should be common to all devices

File: examples/ppo_trainer/run_qwen3_8b_fsdp.sh [hunk 4]
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env bash
+# PPO | text | FSDP training | GPU/NPU
+# Canonical PPO (actor + critic) baseline on GSM8K + MATH.
+
+set -xeuo pipefail
+
+# ---- user-adjustable ----
+DEVICE=${DEVICE:-gpu}
```
Comment (wuxibin89) [inline examples/ppo_trainer/run_qwen3_8b_fsdp.sh]: We can auto detect platform by `import torch_npu` @wucong25

File: examples/grpo_trainer/run_qwen3_8b_fsdp.sh [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    fi
+    n_trainer_devices=${NGPUS_PER_NODE}
+    rollout_tp=${rollout_tp:-2}
+    rollout_gpu_mem_util=${rollout_gpu_mem_util:-0.6}
+    experiment_name=${experiment_name:-qwen3_8b_${INFER_BACKEND}_fsdp${MACHINE:+_${MACHINE}}}
+
+    if [ "${INFER_BACKEND}" = trtllm ]; then
+        optional_ppo_args+=(actor_rollout_ref.hybrid_engine=True)
+    else
+        optional_ppo_args+=(actor_rollout_ref.rollout.mode=async)
```
Comment (wuxibin89) [inline examples/grpo_trainer/run_qwen3_8b_fsdp.sh]: `actor_rollout_ref.rollout.mode=async` and `actor_rollout_ref.hybrid_engine=True` are deprecated and should be removed

File: examples/grpo_trainer/run_qwen3_8b_fsdp.sh [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+    optional_ppo_args+=(
+        actor_rollout_ref.actor.use_torch_compile=False
+        actor_rollout_ref.actor.fsdp_config.ulysses_sequence_parallel_size="${sp_size}"
+        actor_rollout_ref.rollout.enable_chunked_prefill=False
+        actor_rollout_ref.ref.fsdp_config.ulysses_sequence_parallel_size="${sp_size}"
+        trainer.device=npu
+    )
+
+    if [ "${INFER_BACKEND}" = vllm ]; then
```
Comment (wuxibin89) [inline examples/grpo_trainer/run_qwen3_8b_fsdp.sh]: Too much if branch, this should be common config.

File: examples/grpo_trainer/run_qwen3_8b_megatron.sh [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+
+project_name=${PROJECT_NAME:-verl_grpo_gsm8k_math}
+experiment_name=${EXPERIMENT_NAME:-qwen3_8b_${INFER_BACKEND}_megatron}
+########################### end user-adjustable ###########################
+
+########################### derived defaults ###########################
+case "${INFER_BACKEND}" in
+    vllm | sglang | trtllm) ;;
+    *)
+        echo "INFER_BACKEND must be vllm, sglang, or trtllm, got: ${INFER_BACKEND}" >&2
```
Comment (wuxibin89) [inline examples/grpo_trainer/run_qwen3_8b_megatron.sh]: This is not needed in bash script, let it fails in python code.

File: examples/grpo_trainer/run_qwen2_5_vl_7b_fsdp.sh [hunk 8]
```diff
@@ -115,31 +99,19 @@ TRAINER=(
...
+    actor_rollout_ref.actor.strategy=fsdp2
+    actor_rollout_ref.model.use_fused_kernels=True
+    actor_rollout_ref.rollout.multi_stage_wake_up=True
+    actor_rollout_ref.rollout.enable_chunked_prefill=False
+    actor_rollout_ref.rollout.enforce_eager=False
+    actor_rollout_ref.rollout.free_cache_engine=True
+)
+
+if [ "${INFER_BACKEND}" = vllm ]; then
+    EXTRA+=(+actor_rollout_ref.rollout.engine_kwargs.vllm.mm_processor_cache_gb=0)
```
Comment (wuxibin89) [inline examples/grpo_trainer/run_qwen2_5_vl_7b_fsdp.sh]: mm_processor_cache_gb can be eliminated.

File: examples/sglang_multiturn/config/geo3k_multiturn_megatron_grpo.yaml [hunk 9]
```diff
@@ -3,7 +3,8 @@ hydra:
     - file://verl/trainer/config
```
Comment (wuxibin89) [inline examples/sglang_multiturn/config/geo3k_multiturn_megatron_grpo.yaml]: remove `sglang_multiturn` folder

File: examples/ppo_trainer/run_qwen3_8b_fsdp.sh [hunk 10]
```diff
@@ -0,0 +1,139 @@
+#!/usr/bin/env bash
+# PPO | text | FSDP training | GPU/NPU
+# Canonical PPO (actor + critic) baseline on GSM8K + MATH.
+
+set -xeuo pipefail
+
+########################### user-adjustable ###########################
+# DEVICE is auto-detected by probing torch_npu; override only for special cases.
+DEVICE=${DEVICE:-$(python3 -c 'import torch_npu' 2>/dev/null && echo npu || echo gpu)}
```
Comment (wuxibin89) [inline examples/ppo_trainer/run_qwen3_8b_fsdp.sh:9]: import torch_npu is pretty slow, better use:
```
pip3 show torch-npu >/dev/null 2>&1 && echo npu || echo gpu
```

**Review by wuxibin89 — APPROVED** [review]
